smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitorTest.kt

```diff
@@ -5,91 +5,261 @@
+            "import aws.smithy.kotlin.runtime.util.length",
+            "",
+            "val foo = it?.foo",
+            "val length = foo?.length ?: 0",
         )
     }
 
-    private fun testVisit(includeInput: Boolean, path: String, expectedActualName: String, vararg vals: String) {
-        val visitor = KotlinJmespathExpressionVisitor(
-            includeInput,
-            model,
-            provider,
-            inputShape,
-            inputSymbol,
-            outputShape,
-            outputSymbol
+    @Test
+    fun testIndexExpression() {
+        assertUnimplemented("foo[0]")
+    }
+
+    @Test
+    fun testLiteralExpressions() {
+        assertVisit("`\"a string\"`", "string", "val string = \"a string\"")
+        assertVisit("`3.14`", "number", "val number = 3.14")
+        assertVisit("`false`", "bool", "val bool = false")
+        assertVisit("`null`", "null")
+        assertUnimplemented("`[]`")
+        assertUnimplemented("`{}`")
+    }
+
+    @Test
+    fun testMultiSelectHashExpression() {
+        assertUnimplemented("foo.{bar: baz}")
+    }
+
+    @Test
+    fun testMultiSelectListExpression() {
+        assertVisit(
+            "foo[][bar, baz]",
+            "projection",
+            "val foo = it?.foo",
+            "val fooOrEmpty = foo ?: listOf()",
+            "val projection = fooOrEmpty.flatMap {",
+            "    val multiSelect = listOfNotNull(",
+            "        run {",
+            "            val bar = it?.bar",
+            "            bar",
+            "        },",
+            "        run {",
+            "            val baz = it?.baz",
+            "            baz",
+            "        },",
+            "    )",
+            "    multiSelect",
+            "}",
         )
+    }
 
-        val expression = JmespathExpression.parse(path)
-        expression.accept(visitor)
+    @Test
+    fun testNotExpression() {
+        assertUnimplemented("!foo")
+    }
+
+    @Test
+    fun testObjectProjection() {
+        assertVisit(
+            "foo.*.bar",
+            "projection",
+            "val foo = it?.foo",
+            "val fooValues = foo?.values ?: listOf()",
+            "val projection = fooValues.flatMap {",
+            "    val bar = it?.bar",
+            "    listOfNotNull(bar)",
+            "}",
+        )
+    }
+
+    @Test
+    fun testProjectionExpression() {
+        assertVisit(
+            "foo[].bar",
+            "projection",
+            "val foo = it?.foo",
+            "val fooOrEmpty = foo ?: listOf()",
+            "val projection = fooOrEmpty.flatMap {",
+            "    val bar = it?.bar",
+            "    listOfNotNull(bar)",
+            "}",
+        )
+    }
 
+    @Test
+    fun testOrExpression() {
+        assertUnimplemented("foo || bar")
+    }
+
+    @Test
+    fun testSliceExpression() {
+        assertUnimplemented("foo[0:1]")
+    }
+
+    @Test
+    fun testSubExpressions() {
+        assertVisit(
+            "foo.bar.baz",
```

Comment:
suggestion

supplying the parameter names for `path` and `expectedActualName` when calling `assertVisit()` in unit tests would go a long way in lending more readability in these tests IMO

Reply:
Excellent idea!